ducktape/decorate.py

```diff
@@ -0,0 +1,229 @@
+        else:
+            return self.wrapped
+
+    @property
+    def __name__(self):
+        return self.test_method.__name__
+
+    def __len__(self):
+        return len(self.kwargs_list)
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        if self.kwargs_pointer < len(self):
+            current_kwargs = self.kwargs_list[self.kwargs_pointer]
+            self.kwargs_pointer += 1
+            return _inject(**current_kwargs)(self.test_method)
+        else:
+            raise StopIteration
+
+
+def i_to_indices(i, values):
+    """Helper method used by matrix decorator which maps an index to a list of indices.
+
+    n = len(values)
+    0 <= index < len(values[0]) * ... * len(values[n])
+
+    index -> [i_0, ..., i_n] where 0 <= i_k < len(values[k])
+    """
+    num_matrix_elements = reduce(lambda x, y: x * len(y), values, 1)
+    assert 0 <= i < num_matrix_elements
+
+    indices = [0] * len(values)
+    for j in range(len(values)):
+        length_of_j = len(values[j])
+
+        indices[j] = i % length_of_j
+        i /= length_of_j
+        if i == 0:
+            break
+
+    return indices
+
+
+def matrix(**kwargs):
+    """Function decorator used to parametrize its with a matrix of values.
+    Decorating a function or method with @matrix "transforms" it into an iterable object.
+
+    Iterating yields the original function with the specified arguments injected, but there are
+    no ordering guarantees.
+
+    Example::
+
+        @matrix(x=[1, 2], y=[-1, -2])
+        def g(x, y):
+            print "x = %s, y = %s" % (x, y)
+
+        for f in g:
+            f()
+
+        # output:
+        # x = 1, y = -1
+        # x = 1, y = -2
+        # x = 2, y = -1
+        # x = 2, y = -2
+
+        # This is equivalent to the above example:
+        @matrix(x=[1, 2])
+        @matrix(y=[-1, -2])
+        def g(x, y):
+            print "x = %s, y = %s" % (x, y)
+    """
+    vars = []
+    values = []
+    for v in kwargs:
+        vars.append(v)
+        values.append(kwargs[v])
+
+    num_params = 0
+    if len(vars) > 0:
+        num_params = reduce(lambda x, y: x * len(y), values, 1)
+
+    def parametrizer(f):
+
+        kwargs_list = []
+
+        for i in range(num_params):
+            indices = i_to_indices(i, values)
+            kwargs_list.append(
+                {vars[j]: values[j][indices[j]] for j in range(len(indices))})
+
+        if _expandable(f):
+            all_kwargs_list = []
+            for sub_kwargs in f.kwargs_list:
+                for kwargs in kwargs_list:
+                    current_kwargs = kwargs.copy()
+                    current_kwargs.update(sub_kwargs)
+                    all_kwargs_list.append(current_kwargs)
+        else:
+            all_kwargs_list = kwargs_list
+
+        wrapped = TestGenerator(f, all_kwargs_list)
+        mark_as(wrapped, MATRIX)
+        return wrapped
+    return parametrizer
+
+
+def parametrize(**kwargs):
```

Comment:
Apparently I like the British spelling, parameterize. For my personal sanity, I recommend adding an alias, `parameterize = parametrize`.